core/src/main/java/org/axonframework/eventsourcing/eventstore/jdbc/JdbcEventStorageEngine.java

```diff
@@ -261,6 +261,38 @@ protected void storeSnapshot(DomainEventMessage<?> snapshot, Serializer serializ
                                    ))));
     }
 
+    @Override
+    public TrackingToken createHeadToken() {
+        String sql = "SELECT max(" + schema.globalIndexColumn() + ") FROM " + schema.domainEventTable();
+        Long index = transactionManager.fetchInTransaction(
+                () -> executeQuery(getConnection(),
+                                   connection -> connection.prepareStatement(sql),
+                                   resultSet -> resultSet.next() ? resultSet.getObject(1, Long.class) : null,
+                                   e -> new EventStoreException("Failed to get head token")));
+        return Optional.ofNullable(index)
+                       .map(seq -> GapAwareTrackingToken.newInstance(seq, Collections.emptySet()))
+                       .orElse(null);
+    }
+
+    @Override
+    public TrackingToken createTokenAt(Instant dateTime) {
+        String sql = "SELECT min(" + schema.globalIndexColumn() + ") - 1 FROM " + schema.domainEventTable() + " WHERE "
+                + schema.timestampColumn() + " >= ?";
+        Long index = transactionManager.fetchInTransaction(
+                () -> executeQuery(getConnection(),
+                                   connection -> {
+                                       PreparedStatement stmt = connection.prepareStatement(sql);
+                                       stmt.setString(1, formatInstant(dateTime));
+                                       return stmt;
+                                   },
+                                   resultSet -> resultSet.next() ? resultSet.getObject(1, Long.class) : null,
+                                   e -> new EventStoreException(format("Failed to get token at [%s]", dateTime))));
+        if (index == null) {
+            return null;
+        }
+        return GapAwareTrackingToken.newInstance(index, Collections.emptySet());
+    }
```

Comment:
Although I understand we're not taking gaps into account for the create functions here, I do envision problems for the `TrackingToken#covers` functions. That function takes gaps into account, but will now always return false if called as a follow up on one of these functions together with an existing token (which might have) gaps.

Not sure how to cover this though...just wanted to note this.

Reply:
Would like to hear @abuijze opinion on this.